### PR TITLE
[Serve] Remove Serve snapshot

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

The Serve controller snapshots information about the Serve application on every control loop. This used to be consumed by the dashboard through the `/api/snapshot` endpoint to report Serve-specific information. However, this endpoint is no longer used, so this snapshot is no longer necessary.

This change removes the snapshot.

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change removes a test that's no longer necessary.

### CODE REVIEW

**Comment by shrekris-anyscale** [comment]
This change is ready for review, but please don't merge it yet. This change likely affects the wheels built in #38408, which are needed for some manual testing. Once that testing is over, we can merge this change.

File: python/ray/serve/controller.py
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:608]: @edoakes @GeneDer This may be a behavior change, so please check this.

We set this event before `ray.killing` the controller actor. However, as far as I can tell, the only call that `awaits` this event is `graceful_shutdown`. The expected behavior for `graceful_shutdown` is that the caller waits until the controller dies, and the caller receives a `RayActorError`. There seems to be no reason to set this event since `graceful_shutdown` shouldn't return before the controller dies.

I had to remove this because there were a few tests in `test_standalone_3` that were failing (e.g. `test_controller_shutdown_gracefully`). They check that calling `graceful_shutdown` raises a `RayActorError`. These were passing because snapshotting the cluster state gave enough time for `ray.kill` to finish executing before the controller exited the control loop, which raised the `RayActorError` before `graceful_shutdown` could return. However, without snapshotting, the `graceful_shutdown` call was returning before the `RayActorError` got raised.

As further evidence that the tests failed due to a timing issue, I got the test to pass by replacing `self._put_serve_snapshot()` with `time.sleep(0.2)`.

Comment (GeneDer) [reply python/ray/serve/controller.py:608]: Hmm, I'm not sure if I understand the problem 😅 `graceful_shutdown` is called in the client to signal to the controller to shutdown itself with or without the client needing to wait for the process (since it's just setting a flag on the controller and controller will deal with the shutdown sequence in the event loop). `RayActorError` is a byproduct of the client waiting for the `graceful_shutdown` to return, but the actor is already dead right? Maybe instead of dropping this event, we can just drop the `RayActorError` error catching given it now returns faster than ray to kill the controller?

Comment (shrekris-anyscale) [reply python/ray/serve/controller.py:608]: `graceful_shutdown` takes in one argument: [`wait`](https://github.com/ray-project/ray/blob/f22259bbd676a9e433c5ef0887c2a5a32887cfef/python/ray/serve/controller.py#L1019). When `wait` is `True`, the client waits until the controller shuts down, which should raise a `RayActorError` (at least based on how [`test_controller_shutdown_gracefully`](https://github.com/ray-project/ray/blob/f22259bbd676a9e433c5ef0887c2a5a32887cfef/python/ray/serve/tests/test_standalone_3.py#L620-L621) is implemented).

We should clarify the expected behavior here. If `wait` is `True`, should the client wait until the Serve controller is dead? I think the best way to ensure that is to wait until the controller is actually dead and raises a `RayActorError`, instead of returning when the event is set.

Comment (GeneDer) [reply python/ray/serve/controller.py:608]: Oh I got your point! Yep, agreed that way we are sure the client can wait until controller is dead :)

**Review by sihanwang41 — APPROVED** [review]

**Review by GeneDer — APPROVED** [review]
LGTM!

**Comment by shrekris-anyscale** [comment]
The test failures are unrelated:

* `Windows Build & Test` ([`test_cancel` and `test_autoscaler_e2e`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/32768#018a101b-b7de-49bf-bece-b7ab8d685c18/6560-9913)) are flaky on `master`:

Screen Shot 2023-08-22 at 11 46 24 AM

Screen Shot 2023-08-22 at 11 46 10 AM

* `Doctest` ([`workflow:doctest`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/32768#018a1026-db0b-4dcb-96a8-351a434b78e8/1692-2238)) is unrelated.

* `Serve Release Tests` ([`tune_serve_golden_notebook_client_smoke_test`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/32768#018a1026-dafa-47d2-bf93-4ad91bb32c24/3951-4748)) has been failing on `master`:

Screen Shot 2023-08-22 at 11 48 17 AM

**Comment by shrekris-anyscale** [comment]
@edoakes This change is ready to merge.

**Review by alanwguo — APPROVED** [review]

**Review by zcin — APPROVED** [review]

**Pull request merged**
